File: backend/services/csvService.js

const fs = require('fs').promises;
const csvParser = require('csv-parser');
const { stringify } = require('csv-stringify');
const { createCanvas } = require('canvas');
const { createReadStream } = require('fs');

class CSVService {
  /**
   * Check if MIME type is CSV-related
   * @param {string} mimetype - File MIME type
   * @returns {boolean} True if CSV type
   */
  isCSVType(mimetype) {
    return [
      'text/csv',
      'application/vnd.ms-excel',
      'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
    ].includes(mimetype);
  }

  /**
   * Parse CSV file and return paginated data
   * @param {string} csvPath - Path to CSV file
   * @param {number} page - Page number (1-indexed)
   * @param {number} limit - Number of rows per page
   * @returns {Object} Paginated CSV data
   */
  async getPaginatedData(csvPath, page = 1, limit = 20) {
    try {
      // First, get total row count
      const totalRows = await this.getRowCount(csvPath);
      
      if (totalRows === 0) {
        return {
          rows: [],
          totalRows: 0
        };
      }

      // Calculate offset
      const offset = (page - 1) * limit;
      
      if (offset >= totalRows) {
        throw new Error(`Page ${page} exceeds available data`);
      }

      // Read headers first
      const headers = await this.getHeaders(csvPath);
      
      // Read specific rows
      const rows = await this.getRowsRange(csvPath, offset, limit, headers);
      
      return {
        headers,
        rows,
        totalRows: totalRows - 1 // Subtract header row
      };
      
    } catch (error) {
      console.error('❌ CSV pagination failed:', error);
      throw new Error(`Failed to paginate CSV: ${error.message}`);
    }
  }

  /**
   * Count total rows in CSV file
   * @param {string} csvPath - Path to CSV file
   * @returns {number} Total row count including header
   */
  async getRowCount(csvPath) {
    return new Promise((resolve, reject) => {
      let count = 0;
      
      createReadStream(csvPath)
        .pipe(csvParser())
        .on('data', () => count++)
        .on('end', () => resolve(count + 1)) // +1 for header
        .on('error', reject);
    });
  }

  /**
   * Get CSV headers
   * @param {string} csvPath - Path to CSV file
   * @returns {Array} Array of header names
   */
  async getHeaders(csvPath) {
    return new Promise((resolve, reject) => {
      let headers = null;
      
      createReadStream(csvPath)
        .pipe(csvParser())
        .on('headers', (headerList) => {
          headers = headerList;
        })
        .on('data', () => {
          // We only need headers, so we can stop after first row
          if (headers) {
            resolve(headers);
          }
        })
        .on('end', () => {
          resolve(headers || []);
        })
        .on('error', reject);
    });
  }

  /**
   * Get specific range of rows from CSV
   * @param {string} csvPath - Path to CSV file
   * @param {number} offset - Starting row offset
   * @param {number} limit - Number of rows to return
   * @param {Array} headers - CSV headers
   * @returns {Array} Array of row objects
   */
  async getRowsRange(csvPath, offset, limit, headers) {
    return new Promise((resolve, reject) => {
      const rows = [];
      let currentRow = 0;
      
      createReadStream(csvPath)
        .pipe(csvParser())
        .on('data', (row) => {
          if (currentRow >= offset && currentRow < offset + limit) {
            rows.push(row);
          }
          currentRow++;
          
          // Stop reading once we have enough rows
          if (rows.length >= limit) {
            resolve(rows);
          }
        })
        .on('end', () => resolve(rows))
        .on('error', reject);
    });
  }

  /**
   * Generate a visual preview of CSV data as an image
   * @param {string} csvPath - Path to CSV file
   * @param {Object} options - Preview options
   * @returns {Buffer} JPEG image buffer
   */
  async generatePreview(csvPath, options = {}) {
    const {
      maxRows = 6,
      maxCols = 5,
      cellWidth = 120,
      cellHeight = 30,
      fontSize = 12
    } = options;

    try {
      // Get first few rows for preview
      const data = await this.getPaginatedData(csvPath, 1, maxRows);
      
      if (!data.headers || data.headers.length === 0) {
        throw new Error('CSV file appears to be empty or invalid');
      }

      // Limit columns for preview
      const headers = data.headers.slice(0, maxCols);
      const rows = data.rows.map(row => 
        headers.map(header => {
          const value = row[header];
          // Truncate long values
          return typeof value === 'string' && value.length > 15 
            ? value.substring(0, 12) + '...' 
            : value || '';
        })
      );

      // Calculate canvas size
      const canvasWidth = Math.max(600, headers.length * cellWidth + 40);
      const canvasHeight = (rows.length + 2) * cellHeight + 80; // +2 for header and title

      // Create canvas
      const canvas = createCanvas(canvasWidth, canvasHeight);
      const ctx = canvas.getContext('2d');

      // Fill background
      ctx.fillStyle = '#ffffff';
      ctx.fillRect(0, 0, canvasWidth, canvasHeight);

      // Set font
      ctx.font = `${fontSize}px Arial, sans-serif`;
      ctx.textAlign = 'left';
      ctx.textBaseline = 'middle';

      // Draw title
      ctx.fillStyle = '#1f2937';
      ctx.font = 'bold 16px Arial, sans-serif';
      ctx.fillText('CSV Preview', 20, 25);
      
      // Draw table info
      ctx.font = '12px Arial, sans-serif';
      ctx.fillStyle = '#6b7280';
      ctx.fillText(`${data.totalRows} rows × ${data.headers.length} columns`, 20, 45);

      // Draw headers
      const startY = 70;
      ctx.fillStyle = '#f3f4f6';
      ctx.fillRect(20, startY, canvasWidth - 40, cellHeight);
      
      ctx.strokeStyle = '#d1d5db';
      ctx.lineWidth = 1;
      ctx.strokeRect(20, startY, canvasWidth - 40, cellHeight);

      ctx.fillStyle = '#374151';
      ctx.font = `bold ${fontSize}px Arial, sans-serif`;
      
      headers.forEach((header, colIndex) => {
        const x = 30 + colIndex * cellWidth;
        const text = header.length > 15 ? header.substring(0, 12) + '...' : header;
        ctx.fillText(text, x, startY + cellHeight / 2);
        
        // Draw column separators
        if (colIndex > 0) {
          ctx.beginPath();
          ctx.moveTo(x - 10, startY);
          ctx.lineTo(x - 10, startY + cellHeight);
          ctx.stroke();
        }
      });

      // Draw data rows
      ctx.font = `${fontSize}px Arial, sans-serif`;
      rows.forEach((row, rowIndex) => {
        const y = startY + (rowIndex + 1) * cellHeight;
        
        // Alternate row colors
        if (rowIndex % 2 === 0) {
          ctx.fillStyle = '#f9fafb';
          ctx.fillRect(20, y, canvasWidth - 40, cellHeight);
        }

        // Draw row border
        ctx.strokeStyle = '#e5e7eb';
        ctx.strokeRect(20, y, canvasWidth - 40, cellHeight);

        // Draw cell data
        ctx.fillStyle = '#374151';
        row.forEach((cell, colIndex) => {
          const x = 30 + colIndex * cellWidth;
          ctx.fillText(String(cell), x, y + cellHeight / 2);
          
          // Draw column separators
          if (colIndex > 0) {
            ctx.strokeStyle = '#e5e7eb';
            ctx.beginPath();
            ctx.moveTo(x - 10, y);
            ctx.lineTo(x - 10, y + cellHeight);
            ctx.stroke();
          }
        });
      });

      // Add "..." indicator if there are more rows
      if (data.totalRows > maxRows) {
        const lastY = startY + (rows.length + 1) * cellHeight;
        ctx.fillStyle = '#6b7280';
        ctx.font = 'italic 12px Arial, sans-serif';
        ctx.fillText(`... and ${data.totalRows - maxRows} more rows`, 30, lastY + 15);
      }

      // Convert to JPEG
      return canvas.toBuffer('image/jpeg', { quality: 0.9 });
      
    } catch (error) {
      console.error('❌ CSV preview generation failed:', error);
      throw new Error(`Failed to generate CSV preview: ${error.message}`);
    }
  }

  /**
   * Convert CSV data to different formats
   * @param {string} csvPath - Path to CSV file
   * @param {string} format - Output format ('json', 'xml', etc.)
   * @returns {string} Converted data
   */
  async convertFormat(csvPath, format = 'json') {
    try {
      const data = await this.getPaginatedData(csvPath, 1, 1000); // Get first 1000 rows
      
      switch (format.toLowerCase()) {
        case 'json':
          return JSON.stringify({
            headers: data.headers,
            rows: data.rows,
            totalRows: data.totalRows
          }, null, 2);
          
        default:
          throw new Error(`Format ${format} not supported`);
      }
    } catch (error) {
      console.error('❌ CSV format conversion failed:', error);
      throw error;
    }
  }
}

module.exports = new CSVService();

File: backend_python/services/csv_service.py

import csv
import pandas as pd
from pathlib import Path
from PIL import Image, ImageDraw, ImageFont
import io


class CSVService:
    def is_csv_type(self, mimetype):
        """Check if MIME type is CSV-related"""
        return mimetype in [
            'text/csv',
            'application/vnd.ms-excel',
            'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet'
        ]

    async def get_paginated_data(self, csv_path, page=1, limit=20):
        """Parse CSV file and return paginated data"""
        try:
            # First, get total row count
            total_rows = await self.get_row_count(csv_path)
            
            if total_rows == 0:
                return {
                    'rows': [],
                    'totalRows': 0
                }

            # Calculate offset
            offset = (page - 1) * limit
            
            if offset >= total_rows:
                raise Exception(f"Page {page} exceeds available data")

            # Read specific rows using pandas
            df = pd.read_csv(csv_path)
            
            # Get headers
            headers = df.columns.tolist()
            
            # Get paginated rows
            rows = df.iloc[offset:offset + limit].to_dict('records')
            
            return {
                'headers': headers,
                'rows': rows,
                'totalRows': len(df)
            }
            
        except Exception as error:
            print(f"❌ CSV pagination failed: {error}")
            raise Exception(f"Failed to paginate CSV: {str(error)}")

    async def get_row_count(self, csv_path):
        """Count total rows in CSV file"""
        try:
            df = pd.read_csv(csv_path)
            return len(df) + 1  # +1 for header
        except Exception as error:
            return 0

    async def get_headers(self, csv_path):
        """Get CSV headers"""
        try:
            df = pd.read_csv(csv_path, nrows=0)
            return df.columns.tolist()
        except Exception as error:
            return []

    async def get_rows_range(self, csv_path, offset, limit, headers):
        """Get specific range of rows from CSV"""
        try:
            df = pd.read_csv(csv_path)
            rows = df.iloc[offset:offset + limit].to_dict('records')
            return rows
        except Exception as error:
            return []

    async def generate_preview(self, csv_path, options=None):
        """Generate a visual preview of CSV data as an image"""
        if options is None:
            options = {}
        
        max_rows = options.get('maxRows', 6)
        max_cols = options.get('maxCols', 5)
        cell_width = options.get('cellWidth', 120)
        cell_height = options.get('cellHeight', 30)
        font_size = options.get('fontSize', 12)

        try:
            # Get first few rows for preview
            data = await self.get_paginated_data(csv_path, 1, max_rows)
            
            if not data['headers']:
                raise Exception('CSV file appears to be empty or invalid')

            # Limit columns for preview
            headers = data['headers'][:max_cols]
            rows = []
            for row in data['rows']:
                row_data = []
                for header in headers:
                    value = row.get(header, '')
                    # Truncate long values
                    if isinstance(value, str) and len(value) > 15:
                        value = value[:12] + '...'
                    row_data.append(str(value) if value else '')
                rows.append(row_data)

            # Calculate canvas size
            canvas_width = max(600, len(headers) * cell_width + 40)
            canvas_height = (len(rows) + 2) * cell_height + 80  # +2 for header and title

            # Create image
            img = Image.new('RGB', (canvas_width, canvas_height), color='white')
            draw = ImageDraw.Draw(img)

            try:
                font = ImageFont.truetype("arial.ttf", font_size)
                title_font = ImageFont.truetype("arial.ttf", 16)
            except:
                font = ImageFont.load_default()
                title_font = ImageFont.load_default()

            # Draw title
            draw.text((20, 25), 'CSV Preview', fill='black', font=title_font)
            
            # Draw table info
            draw.text((20, 45), f"{data['totalRows']} rows × {len(data['headers'])} columns", fill='gray', font=font)

            # Draw headers
            start_y = 70
            header_rect = (20, start_y, canvas_width - 20, start_y + cell_height)
            draw.rectangle(header_rect, fill='lightgray', outline='black')

            for col_index, header in enumerate(headers):
                x = 30 + col_index * cell_width
                text = header[:15] + '...' if len(header) > 15 else header
                draw.text((x, start_y + cell_height // 2 - font_size // 2), text, fill='black', font=font)
                
                # Draw column separators
                if col_index > 0:
                    line_x = x - 10
                    draw.line([(line_x, start_y), (line_x, start_y + cell_height)], fill='black')

            # Draw data rows
            for row_index, row in enumerate(rows):
                y = start_y + (row_index + 1) * cell_height
                
                # Alternate row colors
                if row_index % 2 == 0:
                    row_rect = (20, y, canvas_width - 20, y + cell_height)
                    draw.rectangle(row_rect, fill='lightgray', outline='black')

                # Draw row border
                row_rect = (20, y, canvas_width - 20, y + cell_height)
                draw.rectangle(row_rect, outline='black', fill=None)

                # Draw cell data
                for col_index, cell in enumerate(row):
                    x = 30 + col_index * cell_width
                    draw.text((x, y + cell_height // 2 - font_size // 2), str(cell), fill='black', font=font)
                    
                    # Draw column separators
                    if col_index > 0:
                        line_x = x - 10
                        draw.line([(line_x, y), (line_x, y + cell_height)], fill='black')

            # Add "..." indicator if there are more rows
            if data['totalRows'] > max_rows:
                last_y = start_y + (len(rows) + 1) * cell_height
                draw.text((30, last_y + 15), f"... and {data['totalRows'] - max_rows} more rows", fill='gray', font=font)

            # Convert to bytes
            img_bytes = io.BytesIO()
            img.save(img_bytes, format='JPEG', quality=90)
            return img_bytes.getvalue()
            
        except Exception as error:
            print(f"❌ CSV preview generation failed: {error}")
            raise Exception(f"Failed to generate CSV preview: {str(error)}")

    async def convert_format(self, csv_path, format_type='json'):
        """Convert CSV data to different formats"""
        try:
            data = await self.get_paginated_data(csv_path, 1, 1000)  # Get first 1000 rows
            
            if format_type.lower() == 'json':
                import json
                return json.dumps({
                    'headers': data['headers'],
                    'rows': data['rows'],
                    'totalRows': data['totalRows']
                }, indent=2)
            else:
                raise Exception(f"Format {format_type} not supported")
        except Exception as error:
            print(f"❌ CSV format conversion failed: {error}")
            raise error
